src/saveload/saveload.cpp

```diff
@@ -1620,6 +1639,135 @@ void SlObject(void *object, const SaveLoadTable &slt)
+ */
+std::vector<SaveLoad> SlTableHeader(const SaveLoadTable &slt)
+{
+	switch (_sl.action) {
+		case SLA_LOAD_CHECK:
+		case SLA_LOAD: {
+			std::vector<SaveLoad> saveloads;
+
+			/* Build a key lookup mapping based on the available fields. */
+			std::map<std::string, const SaveLoad *> key_lookup;
+			for (auto &sld : slt) {
+				key_lookup[sld.name] = &sld;
+			}
+
+			while (true) {
+				static std::string key;
+				SlObjectMember(nullptr, SLEG_SSTR(key, SLE_STR));
+
+				/* Empty field means end-of-header. */
+				if (key == "") break;
+
+				static uint8 var_type;
+				SlObjectMember(nullptr, SLEG_VAR(var_type, SLE_UINT8));
+
+				auto sld_it = key_lookup.find(key);
+				if (sld_it == key_lookup.end()) {
+					DEBUG(sl, 2, "Field %s not found, skipping", key.c_str());
+
+					/* We don't know this field, so insert a SLE_NULL. */
+					if (var_type != SLE_FILE_STRING) {
+						saveloads.push_back(SLE_NULL(SlCalcConvFileLen(var_type)));
+					} else {
+						/* As we do not know the string length yet, we cannot
+						 * use SLE_NULL for strings. Instead, insert a
+						 * SLEG_SSTR to a dummy variable. */
+						static std::string dummy;
+						saveloads.push_back(SLEG_SSTR(dummy, SLE_FILE_STRING | SLE_VAR_NULL));
+					}
+					continue;
+				}
+
+				/* Validate the type of the field. If it is changed, the
+				 * savegame should have been bumped so we know how to do the
+				 * conversion. If this assert() trigger, that clearly didn't
+				 * happen and this is a friendly poke to the developer to bump
+				 * the savegame version and add conversion code. */
+				if (GetVarFileType(sld_it->second->conv) != var_type) {
+					DEBUG(sl, 1, "Field type for %s was expected to be %d but %d was found", key.c_str(), var_type, GetVarFileType(sld_it->second->conv));
+					SlErrorCorrupt("Field type in savegame different than expected");
+				}
+				saveloads.push_back(*sld_it->second);
+			}
+
+			return saveloads;
+		}
+
+		case SLA_SAVE: {
+			/* Automatically calculate the length? */
+			if (_sl.need_length != NL_NONE) {
+				SlSetLength(SlCalcTableHeader(slt));
+				if (_sl.need_length == NL_CALCLENGTH) break;
+			}
+
+			for (auto &sld : slt) {
+				/* Store the field-name. */
+#ifdef OTTD_ASSERT
+				bool res =
+#endif
+				SlObjectMember(const_cast<SaveLoad *>(&sld), SLE_CONDSSTR(SaveLoad, name, SLE_STR, sld.version_from, sld.version_to));
+				/* Make sure we are not storing empty keys. */
+				assert(!res || sld.name != "");
+
+				/* Store the type of this field; sadly, we cannot use SLE_CONDVAR, as it is a few bits from a field. */
+				static uint8 var_type;
+				var_type = GetVarFileType(sld.conv);
+				SlObjectMember(nullptr, SLEG_CONDVAR(var_type, SLE_UINT8, sld.version_from, sld.version_to));
+			}
+
+			/* Add an empty field as end-of-header marker. */
+			static std::string empty = {};
+			SlObjectMember(nullptr, SLEG_SSTR(empty, SLE_STR));
+
+			break;
+		}
+
+		default: NOT_REACHED();
+	}
+
+	return std::vector<SaveLoad>();
```

Comment:
OCD: put this where the "break" is :)

Reply:
There are several breaks in this function. Are you sure your OCD would appreciate that?